project/src/main/java/com/google/sps/data/PlacesScorerImpl.java

```diff
@@ -30,20 +59,51 @@
     }
 
     // Calculates a score for place,
-    // score calculated by the place's rating and driving duration from the user's location.
-    private double calculatePlaceScore(ImmutableMap<Place, Double> durations, Place place) {
+    // score calculated by the place's rating and driving duration to the user's location.
+    private double calculatePlaceScore(ImmutableMap<Place, Long> durations, Place place) {
         return
             RATING_WEIGHT * (place.rating() / MAX_RATING)
             + DURATION_WEIGHT * Math.max(1 - (durations.get(place) / MAX_DURATION_SECONDS), 0);
     }
 
+    // Scores places by their rating only, used in case of errors in durations calculation.
+    private ImmutableMap<Place, Double> scoreByRating(ImmutableList<Place> places) {
+        return places.stream().collect(
+            ImmutableMap.toImmutableMap(place -> place, place -> place.rating() / MAX_RATING));
+    }
+
     // Returns the duration in seconds from each place on places list to the destination
-    private ImmutableMap<Place, Double> getDurations(
-            ImmutableList<Place> places, LatLng destination) {
-        // TODO (M2): This function will call the Distance matrix API to calculate the durations.
-        // Duration are hardcoded to 30 minutes temporarly.
-        Double duration = 1800D;
-        return places.stream()
-            .collect(ImmutableMap.toImmutableMap(place -> place, place -> duration));
+    private ImmutableMap<Place, Long> getDurations(ImmutableList<Place> places, LatLng destination)
+            throws ApiException, InterruptedException, IOException {
+        ImmutableMap.Builder<Place, Long> durations = new ImmutableMap.Builder<>();
+        LatLng[] origins = places.stream()
+            .map(place -> place.location()).toArray(LatLng[]::new);
+        DistanceMatrixApiRequest distanceRequest =
+            DistanceMatrixApi.newRequest(context)
+                .origins(origins)
+                .destinations(destination)
+                .mode(TravelMode.DRIVING);
+        DistanceMatrix distanceMatrix = getDistanceResults(distanceRequest);
+        for (int i = 0; i < places.size(); i++) {
+            durations.put(places.get(i), distanceMatrix.rows[i].elements[0].duration.inSeconds);
```

Comment:
Does the API guarantee that the distance matrix will contain all the elements? And that they're not null?

Reply:
That's a great point.
Rows length will always match the number of origin elements, regardless of whether they're routable, and so is the elements array. 
What happens in case of a problem in finding a route between a place and destination, is that the element will have a status that is different then "OK".
I added this check (and a a test), and for now a problematic place gets the max duration. I think this place should be filtered out, but I added this as a TODO for now because I think there are more urgent things to deal with.

Reply:
👍